README
NAME
    Fatal::Exception - succeed or throw exception

SYNOPSIS
      use Fatal::Exception 'Exception::System' => qw< open close >;
      open my $fh, "/nonexistent";   # throw Exception::System

      use Exception::Base 'Exception::My';
      sub juggle { ... }
      Fatal::Exception->import('Exception::My' => 'juggle');
      juggle;          # succeed or throw exception
      Fatal::Exception->unimport('juggle');
      juggle or die;   # restore original behavior

DESCRIPTION
    Fatal::Exception provides a way to conveniently replace functions which
    normally return a false value when they fail with equivalents which
    raise exceptions if they are not successful. This is the same as Fatal
    module but it throws Exception::Base object on error.

AUTHOR
    Piotr Roszatycki <dexter@debian.org>

LICENSE
    Copyright (C) 2007, 2008 by Piotr Roszatycki <dexter@debian.org>.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See http://www.perl.com/perl/misc/Artistic.html



Changes
------------------------------------------------------------------------
0.03 | Piotr Roszatycki <dexter@debian.org> | 2008-12-21

New:

* Wrapped functions throw Exception::Fatal if died.

  use Fatal::Exception 'Exception::System' => 'open';
  eval { open my $fh, '/notfound', 'r' };
  print ref $@;  # Exception::System
  eval { open my $fh, '/etc/passwd', 'badmode' };
  print ref $@;  # Exception::Fatal

Changes:

* Exception::Argument and Exception::Fatal moved to separate package.
* Build requires Exception::Warning, Exception::Died, Test::Assert.

------------------------------------------------------------------------
0.0204 | Piotr Roszatycki <dexter@debian.org> | 2008-05-08

Changes:

* Requires Exception::Base >= 0.16.

------------------------------------------------------------------------
0.0203 | Piotr Roszatycki <dexter@debian.org> | 2008-04-19

Changes:

* Requires Exception::Base >= 0.15.
* Drop indirect notation is POD and code.

------------------------------------------------------------------------
0.0202 | Piotr Roszatycki <dexter@debian.org> | 2008-04-13

Changes:

* Requires Exception::Base >= 0.14.
* Refactored for new Exception::Base.

------------------------------------------------------------------------
0.0201 | Piotr Roszatycki <dexter@debian.org> | 2007-11-05

Fixes:

* Add Build.PL to MANIFEST.
* Drop dependency on Exception::System.

------------------------------------------------------------------------
0.02 | Piotr Roszatycki <dexter@debian.org> | 2007-10-29

Changes:

* Exception::Fatal is an Exception::Base.
* Ignore one level for exception stack trace.
* Switched to Module::Build.
* Requires Exception::Base >= 0.12.
* Requires Exception::System >= 0.07.
* Requires Test::Unit::Lite >= 0.07.
* Use warnings.

------------------------------------------------------------------------
0.0101 | Piotr Roszatycki <dexter@debian.org> | 2007-06-13

* Minor refactoring.

------------------------------------------------------------------------
0.01 | Piotr Roszatycki <dexter@debian.org> | 2007-05-23

* Initial release.

------------------------------------------------------------------------


t/test.pl
#!/usr/bin/perl

use strict;
use warnings;

use File::Basename;
use File::Spec;
use Cwd;

BEGIN {
    chdir dirname(__FILE__) or die "$!";
    chdir '..' or die "$!";

    unshift @INC, map { /(.*)/; $1 } split(/:/, $ENV{PERL5LIB}) if defined $ENV{PERL5LIB} and ${^TAINT};

    my $cwd = ${^TAINT} ? do { local $_=getcwd; /(.*)/; $1 } : '.';
    unshift @INC, File::Spec->catdir($cwd, 'inc');
    unshift @INC, File::Spec->catdir($cwd, 'lib');
}

use Test::Unit::Lite;

use Exception::Base
    max_arg_nums => 0, max_arg_len => 200, verbosity => 3,
    '+ignore_package' => [ qr/^Test::Unit::/, 'main' ],
    'Fatal::Exception::Warning',
    'Fatal::Exception::Died';

local $SIG{__WARN__} = sub { Fatal::Exception::Warning->throw($_[0], ignore_level => 1) };

all_tests;


t/benchmark.pl
#!/usr/bin/perl -al

package My::Eval;
our $n = 0;
sub test {
    eval { 1; };
    $n++;
}


package My::Ok;
our $n = 0;
sub test {
    eval { opendir F, '.' and close F; };
    $n++;
}


package My::DieScalar;
our $n = 0;
sub test {
    eval { opendir F, '/filenotfound' or die "Message\n"; };
    if ($@ eq "Message\n") { $n++; }
}


package My::DieObject;
our $n = 0;
sub test {
    eval { opendir F, '/filenotfound' or throw My::DieObject };
    if ($@ and $@->isa('My::DieObject')) { $n++; }
}
sub throw {
    my %args = @_;
    die bless {%args}, shift;
}


package My::FatalOk;
our $n = 0;
sub test {
    use Fatal 'opendir';
    eval { opendir F, '.' and close F; };
    $n++;
}


package My::Fatal;
our $n = 0;
sub test {
    use Fatal 'opendir';
    eval { opendir F, '/filenotfound' };
    if ($@ and $@ ne '') { $n++; }
}


package My::ExceptionBase;
use lib '../lib';	
use Exception::Base ':all';
our $n = 0;
sub test {
    try eval { opendir F, '/filenotfound' or Exception::Base->throw(message=>'Message'); };
    if (catch my $e) {
        if ($e->isa('Exception::Base') and $e->with('Message')) { $n++; }
    }
}


package My::ExceptionBase1;
use lib 'lib';	
use Exception::Base ':all';
our $n = 0;
sub test {
    try eval { opendir F, '/filenotfound' or Exception::Base->throw(message=>'Message', verbosity=>1); };
    if (catch my $e) {
        if ($e->isa('Exception::Base') and $e->with('Message')) { $n++; }
    }
}


package My::FatalExceptionOk;
our $n = 0;
sub test {
    use Fatal::Exception 'Exception::Base' => 'opendir';
    eval { open F, '.' and close F; };
    $n++;
}


package My::FatalException;
our $n = 0;
sub test {
    use Exception::Base ':all';
    use Fatal::Exception 'Exception::Base' => 'opendir';
    try eval { opendir F, '/filenotfound' };
    if (catch my $e) {
        if ($e->isa('Exception::Base') and $e->with('Message')) { $n++; }
    }
}


package main;

use Benchmark ':all';

my $result = timethese(-1, {
    '1_Ok'                      => sub { My::Ok::test; },
    '2_DieScalar'               => sub { My::DieScalar::test; },
    '3_DieObject'               => sub { My::DieObject::test; },
    '4_FatalOk'                 => sub { My::FatalOk::test; },
    '5_Fatal'                   => sub { My::Fatal::test; },
    '6_ExceptionBase'           => sub { My::ExceptionBase::test; },
    '7_ExceptionBase1'          => sub { My::ExceptionBase1::test; },
    '8_FatalExceptionOk'        => sub { My::FatalExceptionOk::test; },
    '9_FatalException'          => sub { My::FatalException::test; },
});

cmpthese($result);


t/tlib/Fatal/ExceptionTest.pm
package Fatal::ExceptionTest;

use strict;
use warnings;

use base 'Test::Unit::TestCase';

use Fatal::Exception;

# non-CORE functions from own package
sub sub_test1 {
    return shift();
}

# non-CORE functions outer own package
{
    package Fatal::ExceptionTest::Package1;
    sub sub_test2 {
        return shift();
    }
}

# Should be before import test. Test::Unit can't sort subs' names.
sub test____sane {
    my $self = shift;

    local *FOO;

    my $file = __FILE__;
    eval 'open FOO, "<", "$file"';
    $self->assert_equals('', $@);
    $self->assert_matches(qr/^package/, scalar(<FOO>));

    eval 'close FOO';
    $self->assert_equals('', $@);

    eval 'opendir FOO, "."';
    $self->assert_equals('', $@);

    eval 'close FOO';
    $self->assert_equals('', $@);

    eval 'sub_test1 undef';
    $self->assert_equals('', $@);

    eval 'Fatal::ExceptionTest::Package1::sub_test2 undef';
    $self->assert_equals('', $@);
}

sub test_import {
    my $self = shift;

    local *FOO;

    # empty args
    eval 'Fatal::Exception->import()';
    $self->assert_equals('', $@);
    eval 'Fatal::Exception->unimport()';
    $self->assert_equals('', $@);

    # not enough args
    eval 'Fatal::Exception->import("open")';
    $self->assert_matches(qr/Not enough arguments/, $@);

    # not such exception
    eval 'Fatal::Exception->import("Exception::Fatal::import::NotFound", "open")';
    $self->assert_matches(qr/Cannot find/, $@);

    # not such function
    eval 'Fatal::Exception->import("Exception::Fatal", "notsuchfunction$^T$$")';
    $self->assert_matches(qr/is not a Perl subroutine/, $@);

    # first wrapping
    eval 'use Exception::Base "Exception::Fatal::import::Test1"';
    $self->assert_equals('', $@);
    eval 'Fatal::Exception->import("Exception::Fatal::import::Test1", "open", "sub_test1", "Fatal::ExceptionTest::Package1::sub_test2", ":void", "opendir")';
    $self->assert_equals('', $@);

    my $file = __FILE__;
    eval 'open FOO, "<", "$file"';
    $self->assert_equals('', $@);
    $self->assert_matches(qr/^package/, scalar(<FOO>));

    eval 'close FOO';
    $self->assert_equals('', $@);

    # : too many args
    eval 'open 1, 2, 3, 4, 5';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal', ref $@);

    # : wrapped void=0
    eval 'open FOO, "<", "/doesnotexists$^T$$"';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test1', ref $@);

    # : wrapped void=1 in array context
    eval 'opendir FOO, "/doesnotexists$^T$$"';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test1', ref $@);

    # : wrapped void=1 in scalar context
    eval 'my $ret1 = opendir FOO, "/doesnotexists$^T$$"';
    $self->assert_equals('', $@);

    # : wrapped non-core, our package
    eval 'sub_test1 undef';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test1', ref $@);

    # : wrapped non-core, not our package
    eval 'Fatal::ExceptionTest::Package1::sub_test2 undef';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test1', ref $@);

    # re-wrapping, another exception
    eval 'use Exception::Base "Exception::Fatal::import::Test2"';
    $self->assert_equals('', $@);
    eval 'Fatal::Exception->import("Exception::Fatal::import::Test2", "open", "sub_test1", "Fatal::ExceptionTest::Package1::sub_test2", ":void", "opendir")';
    $self->assert_equals('', $@);

    # : wrapped void=0
    eval 'open FOO, "<", "/doesnotexists$^T$$"';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # : wrapped void=1 in array context
    eval 'opendir FOO, "/doesnotexists$^T$$"';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # : wrapped void=1 in scalar context
    eval 'my $ret1 = opendir FOO, "/doesnotexists$^T$$"';
    $self->assert_equals('', $@);

    # : wrapped non-core, our package
    eval 'sub_test1 undef';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # : wrapped non-core, not our package
    eval 'Fatal::ExceptionTest::Package1::sub_test2 undef';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # re-wrapping, the same exception
    eval 'Fatal::Exception->import("Exception::Fatal::import::Test2", "open", "sub_test1", "Fatal::ExceptionTest::Package1::sub_test2", ":void", "opendir")';
    $self->assert_equals('', $@);

    # : wrapped void=0
    eval 'open FOO, "<", "/doesnotexists$^T$$"';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # : wrapped void=1 in array context
    eval 'opendir FOO, "/doesnotexists$^T$$"';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # : wrapped void=1 in scalar context
    eval 'my $ret1 = opendir FOO, "/doesnotexists$^T$$"';
    $self->assert_equals('', $@);

    # : wrapped non-core, our package
    eval 'sub_test1 undef';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # : wrapped non-core, not our package
    eval 'Fatal::ExceptionTest::Package1::sub_test2 undef';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # un-wrap some functions
    eval 'Fatal::Exception->unimport("open", "sub_test1", ":void", "notexists$^T$$")';

    # : un-wrapped
    eval 'open FOO, "<", "/doesnotexists$^T$$"';
    $self->assert_equals('', $@);

    # : un-wrapped
    eval 'sub_test1 undef';
    $self->assert_equals('', $@);

    # : wrapped
    eval 'opendir FOO, "/doesnotexists$^T$$"';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # : wrapped non-core, not our package
    eval 'Fatal::ExceptionTest::Package1::sub_test2 undef';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # un-wrap un-wrapped
    eval 'Fatal::Exception->unimport("open", "sub_test1", ":void", "notexists$^T$$")';

    # : un-wrapped
    eval 'open FOO, "<", "/doesnotexists$^T$$"';
    $self->assert_equals('', $@);

    # : un-wrapped
    eval 'sub_test1 undef';
    $self->assert_equals('', $@);

    # : wrapped
    eval 'opendir FOO, "/doesnotexists$^T$$"';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # : wrapped non-core, not our package
    eval 'Fatal::ExceptionTest::Package1::sub_test2 undef';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # re-wrapping un-wrapped
    eval 'Fatal::Exception->import("Exception::Fatal::import::Test2", "open", "sub_test1", "Fatal::ExceptionTest::Package1::sub_test2", ":void", "opendir")';
    $self->assert_equals('', $@);

    # : wrapped void=0
    eval 'open FOO, "<", "/doesnotexists$^T$$"';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # : wrapped void=1 in array context
    eval 'opendir FOO, "/doesnotexists$^T$$"';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # : wrapped void=1 in scalar context
    eval 'my $ret1 = opendir FOO, "/doesnotexists$^T$$"';
    $self->assert_equals('', $@);

    # : wrapped non-core, our package
    eval 'sub_test1 undef';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);

    # : wrapped non-core, not our package
    eval 'Fatal::ExceptionTest::Package1::sub_test2 undef';
    $self->assert_not_equals('', $@);
    $self->assert_equals('Exception::Fatal::import::Test2', ref $@);
}

1;


Build.PL
#!/usr/bin/perl

use 5.006;
use Module::Build;

use strict;
use warnings;

my $build = Module::Build->new(
    module_name => 'Fatal::Exception',
    dist_author => 'Piotr Roszatycki <dexter@debian.org>',
    license => 'perl',
    requires => {
        'perl'               => 5.006,
        'Exception::Base'    => 0.21,
    },
    optional => {
    },
    build_requires => {
	'Test::Assert'       => 0,
	'Test::Unit::Lite'   => 0.07,
    },
    create_makefile_pl => 'traditional',
    create_readme => 1,
    recursive_test_files => 1,
    add_to_cleanup => [
        'META.yml', '*.bak', '*.gz', '*.tgz', '*.ppd', '*.deb', '*.bat',
        'Makefile.PL', 'Makefile', 'pm_to_blib', 'README', 'cover_db',
        'Debian_CPANTS.txt',
        'Fatal-Exception-*',
    ],
);

$build->create_build_script;


lib/Exception/Fatal.pm
#!/usr/bin/perl -c

package Exception::Fatal;

=head1 NAME

Exception::Fatal - Thrown when core function has a fatal error

=head1 SYNOPSIS

  use Exception::Fatal;

  eval {
      open my $fh, '/etc/passwd', '+badmode';
  };
  if ($@) {
      my $e = Exception::Fatal->catch;
      $e->throw( message => 'Cannot open', function => 'open' );
  };

=head1 DESCRIPTION

This class is an L<Exception::Died> exception thrown when core function has a
trappable fatal error.

=for readme stop

=cut

use 5.006;
use strict;
use warnings;

our $VERSION = 0.04;


use Exception::Base 0.21 (
    'Exception::Fatal' => {
        isa       => 'Exception::Died',
        has       => 'function',
        message   => 'Unknown function failed',
    },
);


1;


__END__

=begin umlwiki

= Class Diagram =

[                    <<exception>>
                    Exception::Fatal
 ----------------------------------------------------
 +message : Str = "Unknown function failed" {rw, new}
 +function : Str                            {rw, new}
 ----------------------------------------------------
                                                     ]

[Exception::Fatal] ---|> [Exception::Died]

=end umlwiki

=head1 BASE CLASSES

=over

=item *

L<Exception::Died>

=back

=head1 ATTRIBUTES

This class provides new attributes.  See L<Exception::Base> for other
descriptions.

=over

=item message : Str = "Unknown function failed" {rw}

Contains the message of the exception.  This class overrides the default value
from L<Exception::Base> class.

=item function : Str {rw}

Contains the name of function which failed with fatal error.

=back

=head1 SEE ALSO

L<Exception::Died>, L<Exception::Fatal>, L<Fatal::Exception>.

=head1 BUGS

If you find the bug, please report it.

=for readme continue

=head1 AUTHOR

Piotr Roszatycki E<lt>dexter@debian.orgE<gt>

=head1 LICENSE

Copyright (C) 2008 by Piotr Roszatycki E<lt>dexter@debian.orgE<gt>.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>


lib/Fatal/Exception.pm
#!/usr/bin/perl -c

package Fatal::Exception;

=head1 NAME

Fatal::Exception - succeed or throw exception

=head1 SYNOPSIS

  use Fatal::Exception 'Exception::System' => qw< open close >;
  open my $fh, "/nonexistent";   # throw Exception::System

  use Exception::Base 'Exception::My';
  sub juggle { ... }
  Fatal::Exception->import('Exception::My' => 'juggle');
  juggle;          # succeed or throw exception
  Fatal::Exception->unimport('juggle');
  juggle or die;   # restore original behavior

=head1 DESCRIPTION

L<Fatal::Exception> provides a way to conveniently replace functions which
normally return a false value when they fail with equivalents which raise
exceptions if they are not successful.  This is the same as Fatal module but
it throws L<Exception::Base> object on error.

=cut


use 5.006;
use strict;
use warnings;

our $VERSION = 0.04;


use Symbol ();


use Exception::Base (
    '+ignore_package' => __PACKAGE__,
);
use Exception::Argument;
use Exception::Fatal;


# Switch to enable dump for created wrapper functions
our $Debug;


# Cache for not fatalized functions. The key is "$sub".
our %Not_Fatalized_Functions;


# Cache for fatalized functions. The key is "$sub:$exception:$void".
our %Fatalized_Functions;


# Export the wrapped functions to the caller
sub import {
    my $pkg = shift;
    my $exception = shift || return;

    Exception::Argument->throw(
        message => 'Not enough arguments for "' . __PACKAGE__ . '->import"',
        method  => 'import',
    ) unless @_;

    my $mod_version = $exception->VERSION || 0;
    if (not $mod_version) {
        eval "use $exception;";
        if ($@) {
            Exception::Fatal->throw(
                message => "Cannot find \"$exception\" exception class",
            );
        };
    };

    my $callpkg = caller;
    my $void = 0;

    foreach my $arg (@_) {
        if ($arg eq ':void') {
            $void = 1;
        }
        else {
            my $sub = $arg =~ /::/
                      ? $arg
                      : $callpkg . '::' . $arg;
            (my $name = $sub) =~ s/^&?(.*::)?//;

            __make_fatal(
                exception => $exception,
                name      => $name,
                pkg       => $callpkg,
                sub       => $sub,
                void      => $void,
            );
        };
    };

    return 1;
};


# Restore the non fatalized functions to the caller
sub unimport {
    my $pkg = shift;

    my $callpkg = caller;

    foreach my $arg (@_) {
        next if ($arg eq ':void');

        my $sub = $arg =~ /::/
                ? $arg
                : $callpkg . '::' . $arg;
        (my $name = $sub) =~ s/^&?(.*::)?//;

        __make_not_fatal(
            name => $name,
            pkg  => $callpkg,
            sub  => $sub
        );
    };

    return 1;
};


# Create the wrapper. Stolen from Fatal.
sub __make_fatal {
    # args:
    #   exception - exception class name
    #   name - base name of sub
    #   pkg  - current package name
    #   sub  - full name of sub
    #   void - is function called in scalar context?
    my (%args) = @_;

    # check args
    Exception::Argument->throw(
        message  => 'Not enough arguments for "' . __PACKAGE__ . '->__make_fatal"',
        function => '__make_fatal',
    ) if grep { not defined } @args{qw< exception name pkg sub >};

    Exception::Argument->throw(
        message  => 'Bad subroutine name for "' . __PACKAGE__ . '": ' . $args{name},
        function => '__make_fatal',
    ) if not $args{name} =~ /^\w+$/;

    my ($proto, $code_proto, $call, $core, $argvs);
    my $cache_key = "$args{sub}:$args{exception}:" . ($args{void} ? 1 : 0);
    if (defined $Fatalized_Functions{$cache_key} and defined $Not_Fatalized_Functions{$args{sub}}) {
        # already wrapped: restore from cache
        undef *{ Symbol::qualify_to_ref($args{sub}) };
        return *{ Symbol::qualify_to_ref($args{sub}) } = $Fatalized_Functions{$cache_key};
    }
    elsif (defined(&{$args{sub}}) and not eval { prototype "CORE::$args{name}" }) {
        # user subroutine
        $call = "&{\$" . __PACKAGE__ . "::Not_Fatalized_Functions{\"$args{sub}\"}}";
        $proto = prototype $args{sub};
        $Not_Fatalized_Functions{$args{sub}} = \&{$args{sub}}
            unless defined $Not_Fatalized_Functions{$args{sub}};
    }
    else {
        # CORE subroutine
        $core = 1;
        $call = "CORE::$args{name}";
        $proto = eval { prototype $call };

        # not found as CORE subroutine
        Exception::Argument->throw(
            message  => "\"$args{sub}\" is not a Perl subroutine",
            function => '__make_fatal',
        ) unless $proto;

        # create package's function
        if (not defined &{$args{sub}}) {
            # not package's function yet
            $argvs = __fill_argvs($proto);
            my $name = "__$args{name}__Fatal__Exception__not_wrapped";
            my $code = "package $args{pkg};\n"
                     . "sub $name ($proto) {\n"
                     . "    no strict 'refs';\n"
                     .      __write_invocation(
                                %args,
                                argvs => $argvs,
                                call  => $call,
                                orig  => 1,
                            )
                     . "}\n";
            print STDERR $code if $Debug;

            eval $code;
            if ($@) {
                Exception::Fatal->throw(
                    message => "Cannot create \"$args{sub}\" subroutine",
                );
            };

            my $sub = "$args{pkg}::$name";
            print STDERR "*{ $args{sub} } = \\&$sub;\n" if $Debug;
            undef *{ Symbol::qualify_to_ref($args{sub}) };
            *{ Symbol::qualify_to_ref($args{sub}) } = \&$sub;
        };

        if (not defined $Not_Fatalized_Functions{$args{sub}}) {
            $Not_Fatalized_Functions{$args{sub}} = \&{$args{sub}};
        };
    };

    if (defined $proto) {
        $code_proto = " ($proto)";
    }
    else {
        $code_proto = '';
        $proto = '@';
    };

    $argvs = __fill_argvs($proto) if not defined $argvs;

    # define new named subroutine (anonymous would be harder to debug from stacktrace)
    my $name = "__$args{name}__Fatal__Exception__$args{exception}" . ($args{void} ? '_void' : '') . "__wrapped";
    $name =~ tr/:/_/;

    my $code = "package $args{pkg};\n"
             . "sub $name$code_proto {\n"
             . "    no strict 'refs';\n"
             .      __write_invocation(
                        %args,
                        argvs => $argvs,
                        call  => $call,
                    )
             . "}\n";
    print STDERR $code if $Debug;

    my $newsub = eval $code;
    if ($@) {
        Exception::Fatal->throw(
            message => "Cannot create \"$args{sub}\" subroutine",
        );
    };

    my $sub = "$args{pkg}::$name";
    print STDERR "*{ $args{sub} } = \\&$sub;\n" if $Debug;

    undef *{ Symbol::qualify_to_ref($args{sub}) };
    return *{ Symbol::qualify_to_ref($args{sub}) } = $Fatalized_Functions{$cache_key} = \&$sub;
};


# Restore the not-fatalized function.
sub __make_not_fatal {
    # args:
    #   name - base name of sub
    #   pkg  - current package name
    #   sub  - full name of sub
    my (%args) = @_;

    # check args
    Exception::Argument->throw(
        message  => 'Not enough arguments for "' . __PACKAGE__ . '->__make_non_fatal"',
        function => '__make_not_fatal',
    ) if grep { not defined } @args{qw< name pkg sub >};

    Exception::Argument->throw(
        message  => 'Bad subroutine name for "' . __PACKAGE__ . '": ' . $args{name},
        function => '__make_not_fatal',
    ) if not $args{name} =~ /^\w+$/;

    # not wrapped - do nothing
    return unless defined $Not_Fatalized_Functions{$args{sub}};

    undef *{ Symbol::qualify_to_ref($args{sub}) };
    return *{ Symbol::qualify_to_ref($args{sub}) } = $Not_Fatalized_Functions{$args{sub}};
};


# Fill argvs array based on function prototype. Stolen from Fatal.
sub __fill_argvs {
    my $proto = shift;

    my $n = -1;
    my (@code, @protos, $seen_semi);

    while ($proto =~ /\S/) {
        $n++;
        if ($seen_semi) {
            push(@protos,[$n,@code]);
        };
        if ($proto =~ s/^\s*\\([\@%\$\&])//) {
            push(@code, $1 . "{\$_[$n]}");
            next;
        };
        if ($proto =~ s/^\s*([*\$&])//) {
            push(@code, "\$_[$n]");
            next;
        };
        if ($proto =~ s/^\s*(;\s*)?\@//) {
            push(@code, "\@_[$n..\$#_]");
            last;
        };
        if ($proto =~ s/^\s*;//) {
            $seen_semi = 1;
            $n--;
            next;
        };
        Exception::Argument->throw(
            message  => "Unknown prototype letters: \"$proto\"",
            function => '__fill_argvs',
        );
    };
    push @protos, [$n+1, @code];
    return \@protos;
};


# Write subroutine invocation. Stolen from Fatal.
sub __write_invocation {
    # args:
    #   argvs - ref to prototypes stored as array of array of calling arguments
    #   call  - called sub full name
    #   exception - exception class name
    #   name  - base name of sub
    #   orig  - is function called as non-fatalized version?
    #   void  - is function called in scalar context?
    my (%args) = @_;

    # check args
    Exception::Argument->throw(
        message  => 'Not enough arguments for "' . __PACKAGE__ . '->__write_invocation"',
        function => '__write_invocation',
    ) if grep { not defined } @args{qw< argvs call exception name >};

    my @argvs = @{ $args{argvs} };

    my $code;

    if (@argvs == 1) {
        # No optional arguments
        my @argv = @{ $argvs[0] };
        shift @argv;
        $code =
            "    "
            . __one_invocation(
                %args,
                argv => \@argv,
              )
            . ";\n";
    }
    else {
        my $else = "    ";
        my (@out, @argv, $n);
        while (@argvs) {
            @argv = @{shift @argvs};
            $n = shift @argv;
            push @out, "${else}if (\@_ == $n) {\n";
            $else = "    }\n    els";
            push @out,
                "        return "
                . __one_invocation(
                    %args,
                    argv      => \@argv,
                  )
                . ";\n";
        }
        push @out,
            "    };\n"
          . "    Exception::Argument->throw(\n"
          . "        ignore_level => 1,\n"
          . "        message => \"$args{name}: Do not expect to get \" . scalar \@_ . \" arguments\"\n"
          . "    );\n";
        $code = join '', @out;
    };

    return $code;
};


# Write subroutine invocation. Stolen from Fatal.
sub __one_invocation {
    # args:
    #   argv - ref to prototypes stored as array of calling arguments
    #   call - called sub full name
    #   exception - exception class name
    #   name - base name of sub
    #   orig - is function called as non-fatalized version?
    #   void - is function called in scalar context?
    my (%args) = @_;

    # check args
    Exception::Argument->throw(
        message  => 'Not enough arguments for "' . __PACKAGE__ . '->__one_invocation"',
        function => '__one_invocation',
    ) if grep { not defined } @args{qw< argv call exception name >};

    my $argv = join ', ', @{$args{argv}};

    my $code;

    if ($args{orig}) {
        return "$args{call}($argv)";
    }
    elsif ($args{void}) {
        $code = "(defined wantarray)\n"
             . "            ? $args{call}($argv)\n"
             . "            : do {\n"
             . "                  my \$return = eval {\n"
             . "                      $args{call}($argv);\n"
             . "                  };\n"
             . "                  if (\$@) {\n"
             . "                      Exception::Fatal->throw(\n"
             . "                          ignore_level => 1,\n"
             . "                          function     => \"$args{name}\",\n"
             . "                          message      => \"Cannot $args{name}\",\n"
             . "                      );\n"
             . "                  };\n"
             . "                  \$return;\n"
             . "              } || $args{exception}->throw(\n"
             . "                       ignore_level => 1,\n"
             . "                       function     => \"$args{name}\",\n"
             . "                       message      => \"Cannot $args{name}\",\n"
             . "                   )";
    }
    else {
        $code = "$args{call}($argv)\n"
             . "            || $args{exception}->throw(\n"
             . "                   ignore_level => 1,\n"
             . "                   message      => \"Cannot $args{name}\"\n"
             . "               )";
        $code = "(defined wantarray)\n"
             . "            ? do {\n"
             . "                  my \@return = eval {\n"
             . "                      $args{call}($argv);\n"
             . "                  };\n"
             . "                  if (\$@) {\n"
             . "                      Exception::Fatal->throw(\n"
             . "                          ignore_level => 1,\n"
             . "                          function     => \"$args{name}\",\n"
             . "                          message      => \"Cannot $args{name}\",\n"
             . "                      );\n"
             . "                  };\n"
             . "                  \@return;\n"
             . "              } || $args{exception}->throw(\n"
             . "                       ignore_level => 1,\n"
             . "                       function     => \"$args{name}\",\n"
             . "                       message      => \"Cannot $args{name}\",\n"
             . "                   )\n"
             . "            : do {\n"
             . "                  my \$return = eval {\n"
             . "                      $args{call}($argv);\n"
             . "                  };\n"
             . "                  if (\$@) {\n"
             . "                      Exception::Fatal->throw(\n"
             . "                          ignore_level => 1,\n"
             . "                          function     => \"$args{name}\",\n"
             . "                          message      => \"Cannot $args{name}\",\n"
             . "                      );\n"
             . "                  };\n"
             . "                  \$return;\n"
             . "              } || $args{exception}->throw(\n"
             . "                       ignore_level => 1,\n"
             . "                       function     => \"$args{name}\",\n"
             . "                       message      => \"Cannot $args{name}\",\n"
             . "                   )";
    };

    return $code;
};


1;


__END__

=for readme stop

=head1 IMPORTS

=over

=item use Fatal::Exception I<Exception> => I<function>, I<function>, ...

Replaces the original functions with wrappers which provide do-or-throw
equivalents.  You may wrap both user-defined functions and overridable CORE
operators (except exec, system which cannot be expressed via prototypes) in
this way.

If wrapped function occurs fatal error, the error is converted into
L<Exception::Fatal> exception.

If the symbol B<:void> appears in the import list, then functions named
later in that import list raise an exception only when these are called
in void context.

You should not fatalize functions that are called in list context, because
this module tests whether a function has failed by testing the boolean truth
of its return value in scalar context.

If the exception class is not exist, its module is loaded with "use
I<Exception>" automatically.

=item no Fatal::Exception I<function>, I<function>, ...

Restores original functions for user-defined functions or replaces the
functions with do-without-die wrappers for CORE operators.

In fact, the CORE operators cannot be restored, so the non-fatalized
alternative is provided instead.

The functions can be wrapped and unwrapped all the time.

=back

=head1 PERFORMANCE

The L<Fatal::Exception> module was benchmarked with other implementations.
The results are following:

  ---------------------------------------------------------------
  | Module                      | Success       | Failure       |
  ---------------------------------------------------------------
  | eval/die                    |      289616/s |      236308/s |
  ---------------------------------------------------------------
  | Fatal                       |       94627/s |        8967/s |
  ---------------------------------------------------------------
  | Fatal::Exception            |      143479/s |        9677/s |
  ---------------------------------------------------------------

=head1 SEE ALSO

L<Fatal>, L<Exception::Base>, L<Exception::System>

=head1 BUGS

If you find the bug, please report it.

=for readme continue

=head1 AUTHOR

Piotr Roszatycki E<lt>dexter@debian.orgE<gt>

=head1 LICENSE

Copyright (C) 2007, 2008 by Piotr Roszatycki E<lt>dexter@debian.orgE<gt>.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>
